### ISSUE TITLE
Confusion about how the GDPO advantage values in Figure 2 are computed

### ISSUE THREAD

**Issue by FilippoBoni1921** [NVlabs/GDPO#3]
I am trying to reproduce the example of the paper in figure 2. The code is the following:

```
import numpy as np

rewards = [
    [(0,0),(0,0)],
    [(0,0),(0,1)],
    [(0,1),(0,1)],
    [(0,0),(1,1)],
    [(0,1),(0,2)],
    [(0,2),(0,2)],
]

advantages = []

for r in rewards:
    print("----------------------------------------------")
    print(r)
    r0 = [r[0][0],r[1][0]]
    r1 = [r[0][1],r[1][1]]
    print(r0, r1)
    
    mean0 = np.mean(r0)
    std0 = np.std(r0, axis=0,ddof=1) + 1e-9

    mean1 = np.mean(r1)
    std1 = np.std(r1, axis=0,ddof=1) + 1e-9

    print("mean0", mean0)
    print("std0", std0)
    print("mean1", mean1)
    print("std1", std1)
    
    asum1 = (r[0][0]-mean0)/std0 + (r[0][1]-mean1)/std1
    asum2 = (r[1][0]-mean0)/std0 + (r[1][1]-mean1)/std1

    advantages = [asum1, asum2]

    mean = np.mean(advantages)
    std = np.std(advantages, axis=0,ddof=1) + 1e-9

    
    print(advantages)
    print("mean",mean)
    print("std",std)
    print((asum1 - mean)/std)
    print((asum2 - mean)/std)

# [0,0],
# [0,1],
# [1,1],
# [0,2],
# [1,2],
# [2,2],
``` 

but for the case     `[(0,0),(1,1)],` I get as final advantages [-0.70699, +0.70699] while the paper says I should get +-1.414213560373095

**Comment by nbasyl** [comment]
Hi, I ran your code and did get +-1.41421 for [(0,0),(1,1)], could you double check? 

<img width="637" height="209" alt="Image" src="https://github.com/user-attachments/assets/5e5a2ccb-1ea9-42a2-82dc-881663bcc8cb" />

**Comment by nbasyl** [comment]
I realize what cause the confusion here

```
import numpy as np

rewards = [
    [(0,0),(0,0)],
    [(0,0),(0,1)],
    [(0,1),(0,1)],
    [(0,0),(1,1)],
    [(0,1),(0,2)],
    [(0,2),(0,2)],
]

advantages = []

for r in rewards:
    print("----------------------------------------------")
    print(r)
    r0 = [r[0][0],r[1][0]]
    r1 = [r[0][1],r[1][1]]
    print(r0, r1)
    
    mean0 = np.mean(r0)
    std0 = np.std(r0, axis=0,ddof=1) + 1e-9

    mean1 = np.mean(r1)
    std1 = np.std(r1, axis=0,ddof=1) + 1e-9

    print("mean0", mean0)
    print("std0", std0)
    print("mean1", mean1)
    print("std1", std1)
    
    asum1 = (r[0][0]-mean0)/std0 + (r[0][1]-mean1)/std1
    asum2 = (r[1][0]-mean0)/std0 + (r[1][1]-mean1)/std1

    advantages = [asum1, asum2]
```

The final batch-normalization in your implementation is incorrect. You are still normalizing within each group. To reproduce the results from the paper, you only need to run the code up to the advantage computation as shown above. As noted in the paper, since the batch-wise normalization step in GDPO does not change the number of distinct advantage groups, we omit it for clarity.

**Comment by FilippoBoni1921** [comment]
Hi @nbasyl . Thank you very much for the answer.  The last normalization is the application of the formula (6) of the paper. Let me know if I misunderstood something please

**Comment by nbasyl** [comment]
As mentioned in the paper, Equation 6 is excluded from the example in Figure 2 since batch-wise normalization in GDPO does not affect the number of advantage groups. Moreover the final step of your calculation is also incorrect; Assum1 and Assum2 correspond to intra-group normalized rewards 1 and 2 not the normalize result of different question within a batch

**Comment by wdr-RA02** [comment]
Thank you for the wonderful work! I attempted to perform a theoretical calculation using Eqs (4)\~(6), while the case $[0, 2]$ does yield GDPO advantages of $(-1.4142, 1.4142)$ before batch-wise normalization (i.e. without Eq(6)), I found that: 

$$
\begin{eqnarray}
\mu & = & \text{mean}(A_{sum}^0, A_{sum}^1) 
    & = & \frac{A_{sum}^0 + A_{sum}^1}{2}    
    & = & 0
\end{eqnarray}
$$

and

$$
\begin{eqnarray}
\sigma & = & \sqrt{\frac{(A_{sum}^0 - \mu)^2 + (A_{sum}^1 - \mu)^2}{2-1}} 
       & = & \sqrt{(1.414)^2+(1.414)^2}   
       & = & 2
\end{eqnarray}
$$


Therefore, using Eq (6) we have:

$$
\begin{eqnarray}
\hat{A}_{sum}^{0} & = & \frac{A_{sum}^{0} - \mu}{\sigma + \epsilon }
& = & \frac{1.414 - 0}{2} 
& = & 0.707
\end{eqnarray}
$$

and

$$
\begin{eqnarray}
\hat{A}_{sum}^{1} & = & \frac{A_{sum}^{1} - \mu}{\sigma + \epsilon } 
& = & \frac{-1.414 - 0}{2} 
& = & -0.707
\end{eqnarray}
$$

which yields a batch-wise normalized GDPO advantages of $[-0.707, 0.707]$. 
> The $i$ notation is omitted in the calculation above since the mini-batch only contains one item. i.e. $D_{batch} = {[[0,1], [0,1]]}$

With the same steps, I derived the same GDPO post-norm advantage of $[-0.707, 0.707]$ for sample $[0, 1]$

## Batch-wise norm for sample $[0, 1]$ using Eq (6)

Using Eqs (4) and (5), we could obtain the pre-norm advantages of 

$$
\begin{array}{l}
A_{sum}^0  & = & -0.707,  \\
A_{sum}^1 & = & 0.707 
\end{array}
$$

Which aligns with the calculation provided in the paper. According to Eq (6), we obtain: 

$$
\begin{eqnarray}
\mu' & = & \text{mean}(A_{sum}^0, A_{sum}^1) 
    & = & \frac{A_{sum}^0 + A_{sum}^1}{2}    
    & = & 0
\end{eqnarray}
$$

and

$$
\begin{eqnarray}
\sigma' & = & \sqrt{\frac{(A_{sum}^0 - \mu)^2 + (A_{sum}^1 - \mu)^2}{2-1}} 
       & = & \sqrt{(0.707)^2+(0.707)^2}   
       & = & 1
\end{eqnarray}
$$


Therefore, using Eq (6) we have:

$$
\begin{eqnarray}
\hat{A}_{sum}^{0} & = & \frac{A_{sum}^{0} - \mu'}{\sigma' + \epsilon }
& = & \frac{0.707 - 0}{1} 
& = & 0.707
\end{eqnarray}
$$

and

$$
\begin{eqnarray}
\hat{A}_{sum}^{1} & = & \frac{A_{sum}^{1} - \mu'}{\sigma' + \epsilon } 
& = & \frac{-0.707 - 0}{1} 
& = & -0.707
\end{eqnarray}
$$

which yields a batch-wise normalized GDPO advantages of $[-0.707, 0.707]$, which is the same as that of $[0, 2]$. This seems to contradict the claim of "batch-wise normalization in GDPO does not affect the number of advantage groups.". 

I would be very appreciate if you could correct me for any misunderstanding in the calculation described above.

**Comment by nbasyl** [comment]
If you look carefully, the calculation below does not reflect batch statistics. Both 
𝐴_sum^0 and A_sum^1 are tied to the same question, meaning the below operation reduces to another group-wise normalization. As a result, the effect of separating the reward normalization is canceled out. Please note that a batch is supposed to contain multiple different questions.

$$
\begin{eqnarray}
\mu & = & \text{mean}(A_{sum}^0, A_{sum}^1) 
    & = & \frac{A_{sum}^0 + A_{sum}^1}{2}    
    & = & 0
\end{eqnarray}
$$

and

$$
\begin{eqnarray}
\sigma & = & \sqrt{\frac{(A_{sum}^0 - \mu)^2 + (A_{sum}^1 - \mu)^2}{2-1}} 
       & = & \sqrt{(1.414)^2+(1.414)^2}   
       & = & 2
\end{eqnarray}
$$

**Comment by wdr-RA02** [comment]
Thanks for the clarification. 

My original understanding with regards of Eq (6) is that, the normalization is comprised of both batch-wise (represented by $i' \in D_{batch}$) and group-wise (represented by $j' = 1, 2, ..., G$), which looked like a **global-wise** normalization. Building upon the concept, I mistakenly applied an additional step of normalization over $[A_{sum}^0, A_{sum}^1]$.

Now as I understand, for this specific scenario $[[0, 1], [0, 1]]$, the batch-wise normalization outlined in Eq (6) should be omitted, because in this case, the batch-wise operation would be equivalent to an additional group-wise normalization. 

--------------------------------------------
Just for verification, for another mini-batch with two questions and two rollouts denoted as: 

$$
\mathbf{A}= \left(
\begin{array}{c}
A_{sum}^{(0, 0)} & A_{sum}^{(0, 1)}  \\
A_{sum}^{(1, 0)} & A_{sum}^{(1, 1)}  \\
\end{array}
\right)
$$ 

I assume the normalization batch would be: 

$$
[A_{sum}^{(i', j')} | i' \in D_{batch}, j' = 1, 2, ..., G] = [A_{sum}^{(0, 0)}, A_{sum}^{(0, 1)}, A_{sum}^{(1, 0)}, A_{sum}^{(1, 1)}]
$$

> I apologize for using square brackets instead of braces to cover the collection, as the notion "\\{" would not be rendered correctly T^T

Therefore the $\text{mean}$ and $\text{std}$ operation would be carried out on all four elements (2 questions * 2 rollouts).

Please correct me if I got it wrong again.

**Comment by nbasyl** [comment]
Not entirely correct, the batch-norm is conducted after you sum up the decoupled group-wise advantages, so in your case, you should sum up A^(0,0) + A^(0,1) and A^(1,0) + A^(1,1), then calculate the mean and std of those two values.  Please note that the final batch normalization does not affect the number of distinct advantage groups across different training configurations (e.g., varying numbers of rewards or rollouts). Moreover, the number of distinct advantage groups is independent of the batch size.

**Comment by nbasyl** [comment]
Please also keep this issue open so others with similar questions can benefit.

**Comment by rysaya** [comment]
I have a small question as well. When applying batch-wise advantages normalization in the paper, why do we need to subtract the mean of the batch global advantage? Isn't the mean of the batch global advantage always zero?

**Comment by Zenki229** [comment]
I simply reproduce Figure 3(with the help of Gemini 3 pro) in the following code. I understood that the difference between GRPO and GDPO is the computational order of reward group `reward=[batch_size, n_rollouts, n_rewards]`.

GRPO firstly **compute the stats(mean&std) of reward in axis 2**, while GDPO focus on **axis 1**. 


<img width="1882" height="906" alt="Image" src="https://github.com/user-attachments/assets/36475e6b-f00a-486c-a030-e7870bf63785" />


```
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# 1. Set random seed
np.random.seed(42)

def compute_distinct_groups(n_rollouts, n_rewards, n_samples=50000):
    """
    Simulate and compute the number of Distinct Advantage Groups under four configurations
    """
    # Generate data: [Batch, G, n_rewards]
    rewards = np.random.randint(0, 2, size=(n_samples, n_rollouts, n_rewards))
    
   
    def batch_normalize_tensor(arr, use_std=True):
        # Normalize across the entire Batch dimension
        mean = np.mean(arr) # Global mean
        if use_std:
            std = np.std(arr, ddof=1) # Global standard deviation
            if std < 1e-9: std = 1.0
            return (arr - mean) / std
        else:
            return arr - mean

    def group_normalize(arr, use_std=True):
        # Group normalization (axis=1)
        mean = np.mean(arr, axis=1, keepdims=True)
        if use_std:
            std = np.std(arr, axis=1, keepdims=True, ddof=1)
            std[std < 1e-9] = 1.0
            return (arr - mean) / std
        else:
            return arr - mean

    # === 1. GRPO (Sum -> Normalize) ===
    grpo_sum = np.sum(rewards, axis=2)
    grpo_adv = group_normalize(grpo_sum, use_std=True)
    
    # === 2. GRPO w/o std (Sum -> Normalize without std) ===
    grpo_no_std_adv = group_normalize(grpo_sum, use_std=False)
    
    # === 3. GDPO (Normalize per reward -> Sum) ===
    # First normalize each reward dimension
    means = np.mean(rewards, axis=1, keepdims=True)
    stds = np.std(rewards, axis=1, keepdims=True, ddof=1)
    stds[stds < 1e-9] = 1.0
    gdpo_normed = (rewards - means) / stds
    # Sum
    gdpo_raw_adv = np.sum(gdpo_normed, axis=2)
    
    # === 4. GDPO + Batch Norm (paper's complete version) ===
    # After summing, apply another Batch Normalization to the result (Eq. 6)
    gdpo_bn_adv = batch_normalize_tensor(gdpo_raw_adv, use_std=True)

    # === Distinct Count ===
    def count_unique(adv_batch):
        # Keep 4 decimal places to ignore floating point tiny differences
        adv_rounded = np.round(adv_batch, decimals=4)
        # Sort to eliminate order effects
        adv_sorted = np.sort(adv_rounded, axis=1)
        # Count unique rows
        unique_rows = np.unique(adv_sorted, axis=0)
        return len(unique_rows)

    return {
        "GRPO": count_unique(grpo_adv),
        "GRPO w/o std": count_unique(grpo_no_std_adv),
        "GDPO": count_unique(gdpo_raw_adv),
        "GDPO + BN": count_unique(gdpo_bn_adv) # Newly added item
    }
# ==================== Run Experiments ====================
print("Computing data, please wait...")

# Settings
rollout_range = np.arange(2, 17)
reward_range = np.arange(2, 9)

# Store results
left_data = {k: [] for k in ["GRPO", "GRPO w/o std", "GDPO", "GDPO + BN"]}
right_data = {k: [] for k in ["GRPO", "GRPO w/o std", "GDPO", "GDPO + BN"]}

# Run left plot experiment (Varying Rollouts)
for G in rollout_range:
    res = compute_distinct_groups(n_rollouts=G, n_rewards=2, n_samples=20000)
    for k in left_data: left_data[k].append(res[k])

# Run right plot experiment (Varying Rewards)
for n in reward_range:
    res = compute_distinct_groups(n_rollouts=4, n_rewards=n, n_samples=40000)
    for k in right_data: right_data[k].append(res[k])

# ==================== Plotting (Optimized Legend Layout) ====================

fig = make_subplots(rows=1, cols=2, 
                    subplot_titles=("<b>(a) Varying Number of Rollouts</b><br>(Rewards=2)", 
                                    "<b>(b) Varying Number of Rewards</b><br>(Rollouts=4)"),
                    horizontal_spacing=0.15)

# Style definitions
styles = {
    "GRPO":         dict(color="grey", width=2, dash="solid", marker="circle"),
    "GRPO w/o std": dict(color="#4682B4", width=2, dash="solid", marker="square"),
    "GDPO":         dict(color="green", width=4, dash="solid", marker="triangle-up"),
    "GDPO + BN":    dict(color="orange", width=2, dash="dot", marker="x"),
}

for method in styles:
    fig.add_trace(
        go.Scatter(x=rollout_range, y=left_data[method], 
                   mode='lines+markers', name=method,
                   line=dict(color=styles[method]["color"], width=styles[method]["width"], dash=styles[method]["dash"]),
                   marker=dict(symbol=styles[method]["marker"], size=7), # Slightly larger points
                   legendgroup=method), # This line ensures both plots share one Legend
        row=1, col=1
    )

for method in styles:
    fig.add_trace(
        go.Scatter(x=reward_range, y=right_data[method], 
                   mode='lines+markers', name=method,
                   line=dict(color=styles[method]["color"], width=styles[method]["width"], dash=styles[method]["dash"]),
                   marker=dict(symbol=styles[method]["marker"], size=7),
                   legendgroup=method, showlegend=False), # showlegend=False removes duplicate legend
        row=1, col=2
    )

# ==================== Key Modifications: Layout Settings ====================
fig.update_layout(
    title_text="<b>Reproduction of GDPO Figure 3</b>",
    title_x=0.5,
    title_y=0.95, # Title position
    title_font=dict(size=20),
    
    template="plotly_white",
    height=600, width=1100,
    
    # 1. Increase top margin to prevent legend and title overlap
    margin=dict(t=160, b=60, l=60, r=60),
    
    # 2. Precisely position legend in the margin area
    legend=dict(
        orientation="h",         
        yanchor="bottom", y=1.10, 
        xanchor="center", x=0.5,  
        bgcolor="rgba(255,255,255,1)", # White background to cover
        bordercolor="Black", borderwidth=1,
        font=dict(size=12),
        itemwidth=30 # Control spacing between legend items
    )
)

# Common axis settings
axis_config = dict(showgrid=True, gridwidth=1, gridcolor='LightGrey', 
                   mirror=True, showline=True, linecolor='black', linewidth=1)

fig.update_xaxes(title_text="Number of Rollouts", tickmode='linear', row=1, col=1, **axis_config)
fig.update_yaxes(title_text="Distinct Advantage Group Count", row=1, col=1, **axis_config)

fig.update_xaxes(title_text="Number of Rewards", tickmode='linear', row=1, col=2, **axis_config)
fig.update_yaxes(title_text="Distinct Advantage Group Count", row=1, col=2, **axis_config)

# fig.show()

fig.write_image('./gdpo_reproduce.jpg')
```
